File: iOS/iOSDemo/iOSDemo/LoginOperation/LoginOpHelper.m
//
//  LoginOpHandler.m
//  iOSDemo
//
//  Created by willyy on 2018/5/30.
//  Copyright © 2018年 willyy. All rights reserved.
//

#import "LoginOpHelper.h"
#import "GlobalEnv.h"
#import <RSAppSDK/RSunKit.h>
@implementation LoginOpHelper

/**
 * 设计一个token 格式 以便登陆业务可以使用
 * token = ios+phoneNumber+uuid+deviceType+appName+appVersion
 */
+(NSString*)genLoginTokenSeed:(NSString *)phone{
    NSString *token = @"";
    NSString *uuid = [GlobalEnv getUUID];
    NSString *deviceType = [GlobalEnv getDeviceType];
    NSString *appName = [GlobalEnv getAppName];
    NSString *appVersion = [GlobalEnv getAppVersion];
    token = [@"iOS" stringByAppendingFormat:@"_%@_%@_%@_%@_%@",phone,uuid,deviceType,appName,appVersion];
    return token;
}

+(NSString *)getSaltPasswd:(NSString *)passwd{
//    return [GlobalEnv salt:[GlobalEnv getUUID] passwd:passwd];
    return [RSunKit hash_passwd:passwd salt:[GlobalEnv getUUID]];
}

-(void)getChallenge:(NSString *)token handler:(void (^)(NSDictionary *res,NSError *error))handler{
    NSDictionary *para = @{@"token":token};
    [super network_base:@"/demo/challenge" param:para method:@"GET" progress:nil handler:^(NSDictionary * _Nullable res, NSError * _Nullable error) {
        handler(res,error);
    }];
}

/**
 * 登陆属于低安全场景 不用设置密码, 这样的好处再次打开app时不用输入登陆密码
 */
-(void)getLoginKey:(NSString *)token authCode:(NSString *)authCode handler:(void (^)(NSDictionary *res,NSError *error))handler{
    NSDictionary *param = @{@"token":token,@"authCode":authCode};
    [super network_base:@"/demo/generateKey" param:param method:@"POST" progress:nil handler:^(NSDictionary * _Nullable res, NSError * _Nullable error) {
        handler(res,error);
    }];
}

/**
 * 高安全级时使用的Key
 * 比如付费等操作前需要的鉴权工作
 */
-(void)getPrivilegeKey:(NSString *)token passwd:(NSString *)passwd authCode:(NSString *)authCode handler:(void (^)(NSDictionary *res,NSError *error))handler{
    NSDictionary *param;
    if (passwd==nil) {
       param = @{@"token":token,@"passwd":passwd,@"authCode":authCode};
    }
    else{
    param = @{@"token":token,@"passwd":passwd,@"authCode":authCode,@"passwd":passwd};
    }
    [super network_base:@"/demo/generateKey" param:param method:@"POST" progress:nil handler:^(NSDictionary * _Nullable res, NSError * _Nullable error) {
        handler(res,error);
    }];
}

/**
 * verifySign接口
 */
-(void)verifySign:(NSString *)token sign:(NSString *)sign passwd:(NSString *)passwd handler:(void (^)(NSDictionary *res,NSError *error))handler{
    if (passwd==nil) {
        passwd = @"";
    }
    NSDictionary *param = @{@"token":token, @"sign":sign, @"passwd":passwd};
    [super network_base:@"/demo/verify" param:param method:@"POST" progress:nil handler:^(NSDictionary * _Nullable res, NSError * _Nullable error) {
        handler(res,error);
    }];
}

-(void)getAuthCode:(NSString *)phone handler:(void (^)(NSDictionary *res, NSError *error))handler{
    [super network_base:@"/demo/authCode" param:@{@"phone":phone} method:@"GET" progress:nil handler:^(NSDictionary * _Nullable res, NSError * _Nullable error) {
        handler(res, error);
    }];
}

/**
 * register接口, 业务服务器同时生成低安全级别和高安全级别的Key
 */
-(void)registerUser:(NSString *)phone passwd:(NSString *)passwd authCode:(NSString *)authCode handler:(void (^)(BOOL success, NSString *error))handler{
    NSString *TokenSeed = [[self class] genLoginTokenSeed:phone];
    [super network_base:@"/demo/getLoginToken" param:@{@"seed":TokenSeed} method:@"GET" progress:nil handler:^(NSDictionary * _Nullable res, NSError * _Nullable error) {
        NSString *token = res[@"token"];
        [[GlobalEnv env] setLoginToken:token];
        [super network_base:@"/demo/registerUser" param:@{@"phone":phone,@"token":token,@"passwd":passwd,@"authCode":authCode} method:@"POST" progress:nil handler:^(NSDictionary * _Nullable res, NSError * _Nullable error) {
            if (error!=nil) {
                handler(NO,error.description);
            }
            NSString *err = res[@"error"];
            NSString *lowKey = res[@"low"];
            NSString *highKey = res[@"high"];
            if (err.length == 0) {
                [[GlobalEnv env] setLoginKey:lowKey];
                [[GlobalEnv env] setPrivilegeKey:highKey];
                handler(YES,nil);
            }
            else{
                handler(NO,err);
            }
        }];
    }];
}

@end

File: Android/app/src/main/java/com/rsun/andemo/Operation/LoginOpHelper.java
package com.rsun.andemo.Operation;

import android.content.Context;
import android.util.Log;

import com.rsun.Global.GlobalEnv;
import com.rsun.andemo.inf.RSunApi;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LoginOpHelper {
    private static final String TAG = "LoginOpHelper";

    private static LoginOpHelper mHelper = null;

    private Context mContext;

    private LoginOpHelper(Context context) {
        mContext = context;
    }

    public static LoginOpHelper getHelper(Context context) {
        if (mHelper == null) {
            mHelper = new LoginOpHelper(context);
        }

        return mHelper;
    }

    /**
     * 设计一个token 格式 以便登陆业务可以使用
     * token = android+phoneNumber+uuid+Brand+OSVersion+appName+appVersion+expireTime
     */
    public static String genLoginTokenSeed(String phone){
        String token = "";
        if (GlobalEnv.GUID.length()!=0){
            token = "android"+"_"+phone+"_"+ GlobalEnv.GUID+"_"+GlobalEnv.phoneBrand+"_"+GlobalEnv.OSVersion+"_"+GlobalEnv.packageName+"_"+GlobalEnv.appVersion;
        }
        return token;
    }
    public static String getSaltPasswd(String passwd){
        return GlobalEnv.salt(GlobalEnv.GUID,passwd);
    }

    public interface LoginOpCallback{
        void onResponse(boolean response);
        void onFailure(Error error);
        void challenge(String challenge);
    }

    public void getAuthCode(String phone, final LoginOpCallback callback){
        Call<ResponseBody> challenge = GlobalEnv.getInstance(mContext).getRetrofit().create(RSunApi.class).getAuthCode(phone);
        challenge.enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {

                String t;
                try {
                    t = response.body().string();
                    try {
                        JSONObject jsonResponse = new JSONObject(t);
                        boolean codeSend = jsonResponse.getBoolean("codeSend");
                        callback.onResponse(codeSend);
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }

            }
            @Override
            public void onFailure(Call<ResponseBody> call, Throwable throwable) {
                callback.onFailure(new Error("连接失败"));
            }
        });
    }

    public void registerUser(final String phone, final String passwd, final String authCode, final LoginOpCallback callback){
        String TokenSeed = genLoginTokenSeed(phone);
        if (TokenSeed.length()==0){
            Log.d(TAG, "registerUser: fail to generate token seed");
            return;
        }
        Call<ResponseBody> getLoginTokenQueue = GlobalEnv.getInstance(mContext).getRetrofit().create(RSunApi.class).getLoginToken(TokenSeed);
        getLoginTokenQueue.enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                String t;
                try {
                    t = response.body().string();
                    try {
                        JSONObject jsonResponse = new JSONObject(t);
                        String token = jsonResponse.getString("token");
                        GlobalEnv.getInstance(mContext).setLoginToken(token);
                        Call<ResponseBody> registerUserQueue = GlobalEnv.getInstance(mContext).getRetrofit().create(RSunApi.class).registerUser(phone,token,passwd,authCode);
                        registerUserQueue.enqueue(new Callback<ResponseBody>() {
                            @Override
                            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                                String r;
                                try {
                                    r = response.body().string();
                                    try {
                                        JSONObject jsonObject = new JSONObject(r);
                                        String lowKey = jsonObject.getString("low");
                                        String highKey = jsonObject.getString("high");
                                        GlobalEnv.getInstance(mContext).setLoginKey(lowKey);
                                        GlobalEnv.getInstance(mContext).setPrivilegeKey(highKey);
                                        callback.onResponse(true);
                                    } catch (JSONException e) {
                                        e.printStackTrace();
                                    }
                                } catch (IOException e) {
                                    e.printStackTrace();
                                }
                            }

                            @Override
                            public void onFailure(Call<ResponseBody> call, Throwable t) {
                                callback.onFailure(new Error("连接失败"));
                            }
                        });
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void onFailure(Call<ResponseBody> call, Throwable t) {
                callback.onFailure(new Error("连接失败"));
            }
        });
    }

    public void getChallenge(final String token, final LoginOpCallback callback){
        Call<ResponseBody> getChallengeQueue = GlobalEnv.getInstance(mContext).getRetrofit().create(RSunApi.class).getChallenge(token);
        getChallengeQueue.enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                String t;
                try {
                    t = response.body().string();
                    try {
                        JSONObject jsonResponse = new JSONObject(t);
                        String rand = jsonResponse.getString("rand");
                        callback.challenge(rand);
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void onFailure(Call<ResponseBody> call, Throwable t) {
                callback.onFailure(new Error("连接失败"));
            }
        });
    }

    public void verifySign(final String token, final String sign, final String passwd, final LoginOpCallback callback){
        String pwd;
        if (passwd==null){
            pwd = "";
        }
        else{
            pwd = passwd;
        }
        Call<ResponseBody> verifySignQueue = GlobalEnv.getInstance(mContext).getRetrofit().create(RSunApi.class).verifySign(token,sign,pwd);
        verifySignQueue.enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                String t;
                try {
                    t = response.body().string();
                    try {
                        JSONObject jsonResponse = new JSONObject(t);
                        boolean verify = jsonResponse.getBoolean("verify");
                        callback.onResponse(verify);
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void onFailure(Call<ResponseBody> call, Throwable t) {
                callback.onFailure(new Error("连接失败"));
            }
        });
    }
}
